src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,285 @@
+        self._episode_qf_losses = []
+        self._epoch_ys = []
+        self._epoch_qs = []
+
+        self._policy = policy
+        self._qf = qf
+        self._n_train_steps = n_train_steps
+
+        self._min_buffer_size = min_buffer_size
+        self._qf = qf
+        self._steps_per_epoch = steps_per_epoch
+        self._n_train_steps = n_train_steps
+        self._buffer_batch_size = buffer_batch_size
+        self._discount = discount
+        self._reward_scale = reward_scale
+        self.max_episode_length = env_spec.max_episode_length
+        self._max_episode_length_eval = (max_episode_length_eval
+                                         or self.max_episode_length)
+        self._episode_reward_mean = collections.deque(maxlen=100)
+        self._num_eval_episodes = num_eval_episodes
+        self._deterministic_eval = deterministic_eval
+
+        self.env_spec = env_spec
+        self.replay_buffer = replay_buffer
+        self.policy = policy
+        self.exploration_policy = exploration_policy
+
+        self._target_qf = copy.deepcopy(self._qf)
+        self._qf_optimizer = make_optimizer(qf_optimizer,
+                                            module=self._qf,
+                                            lr=qf_lr)
+        self._eval_env = eval_env
+
+    def train(self, trainer):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            trainer (Trainer): Experiment trainer.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        if not self._eval_env:
+            self._eval_env = trainer.get_env_copy()
+        last_returns = [float('nan')]
+        trainer.enable_logging = False
+
+        for _ in trainer.step_epochs():
+            if (self.replay_buffer.n_transitions_stored >=
+                    self._min_buffer_size):
+                trainer.enable_logging = True
+                logger.log('Evaluating policy')
+
+                params_before = self.exploration_policy.get_param_values()
+                eval_eps = obtain_evaluation_episodes(
+                    (self.exploration_policy
+                     if not self._deterministic_eval else self.policy),
+                    self._eval_env,
+                    num_eps=self._num_eval_episodes,
+                    max_episode_length=self._max_episode_length_eval)
+                self.exploration_policy.set_param_values(params_before)
+
+                last_returns = log_performance(trainer.step_itr,
+                                               eval_eps,
+                                               discount=self._discount)
+                self._episode_reward_mean.extend(last_returns)
+                tabular.record('Evaluation/100EpRewardMean',
+                               np.mean(self._episode_reward_mean))
+
+            for _ in range(self._steps_per_epoch):
+                trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
```

Comment:
over here you can pass the number of timesteps that you want to trainer.obtain_episodes. instead of enabling and disabling logging because you're filling the replay buffer, instead get the number of steps that you need in the first training epoch here.

Reply:
this makes sense to me as the solution most consistent with the sampler API. for instance it allows a parallel sampler to farm out the initial sampling efficiently to workers.

Reply:
just curious, what would be the benefit of doing this over disabling logging? I imagine it would have to be something like:

```python
if self._min_buffer_size - self.replay_buffer.n_transitions_stored > 0:
      trainer.step_path = trainer.obtain_episodes(trainer.step_itr, self._min_buffer_size - self.replay_buffer.n_transitions_stored)
else:
      trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
```

Seems simpler to just disable and re-enable logging?

Reply:
Conceptually it seems a little simpler to everyone else to just call the sampler once up-front, because disable logging is kinda weird.

Reply:
Sounds good.